api/client/client_test.go

```diff
@@ -0,0 +1,502 @@
+	checkErr(t, err)
+
+	txn, err := client.WriteTransaction(storeID, modelName)
+	if err != nil {
+		t.Fatalf("failed to create write txn: %v", err)
+	}
+
+	end, err := txn.Start()
+	defer func() {
+		err = end()
+		if err != nil {
+			t.Fatalf("failed to end txn: %v", err)
+		}
+	}()
+	if err != nil {
+		t.Fatalf("failed to start write txn: %v", err)
+	}
+
+	person := createPerson()
+
+	err = txn.Create(person)
+	if err != nil {
+		t.Fatalf("failed to create in write txn: %v", err)
+	}
+	if person.ID == "" {
+		t.Fatalf("expected an entity id to be set but it wasn't")
+	}
+
+	has, err := txn.Has(existingPerson.ID)
+	if err != nil {
+		t.Fatalf("failed to write txn has: %v", err)
+	}
+	if !has {
+		t.Fatalf("expected has to be true but it wasn't")
+	}
+
+	foundExistingPerson := &Person{}
+	err = txn.FindByID(existingPerson.ID, foundExistingPerson)
+	if err != nil {
+		t.Fatalf("failed to txn find by id: %v", err)
+	}
+	if !reflect.DeepEqual(foundExistingPerson, existingPerson) {
+		t.Fatalf("txn model found by id does't equal the original")
+	}
+
+	existingPerson.Age = 99
+	err = txn.Save(existingPerson)
+	if err != nil {
+		t.Fatalf("failed to save in write txn: %v", err)
+	}
+
+	err = txn.Delete(existingPerson.ID)
+	if err != nil {
+		t.Fatalf("failed to delete in write txn: %v", err)
+	}
+}
+
+func TestListen(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, err := client.NewStore()
+	checkErr(t, err)
+	err = client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	checkErr(t, err)
+
+	person := createPerson()
+
+	err = client.ModelCreate(storeID, modelName, person)
+	checkErr(t, err)
+
+	channel, discard, err := client.Listen(storeID, modelName, person.ID)
+	defer discard()
+	if err != nil {
+		t.Fatalf("failed to listen: %v", err)
+	}
+
+	go func() {
+		time.Sleep(1 * time.Second)
+		person.Age = 30
+		_ = client.ModelSave(storeID, modelName, person)
+		person.Age = 40
+		_ = client.ModelSave(storeID, modelName, person)
+	}()
+
+	val, ok := <-channel
+	if !ok {
+		t.Fatal("channel no longer active at first event")
+	} else {
+		if val.err != nil {
+			t.Fatalf("failed to receive first listen result: %v", val.err)
+		}
+		p := &Person{}
+		if err := json.Unmarshal(val.data, p); err != nil {
+			t.Fatalf("failed to unmarshal listen result: %v", err)
+		}
+		if p.Age != 30 {
+			t.Fatalf("expected listen result age = 30 but got: %v", p.Age)
+		}
+	}
+
+	val, ok = <-channel
+	if !ok {
+		t.Fatal("channel no longer active at second event")
+	} else {
+		if val.err != nil {
+			t.Fatalf("failed to receive second listen result: %v", val.err)
+		}
+		p := &Person{}
+		if err := json.Unmarshal(val.data, p); err != nil {
+			t.Fatalf("failed to unmarshal listen result: %v", err)
+		}
+		if p.Age != 40 {
+			t.Fatalf("expected listen result age = 40 but got: %v", p.Age)
+		}
+	}
+}
+
+func server(t *testing.T) (*api.Server, func()) {
```

Comment:
@jsign now that your recent PR is merged in here, what was the change that you told me we can revert here?

Reply:
Hmm, it was moving the call to `server.Close()` in the returned cleanup method to after the call to `ts.Close()`, but I tried reverting and none of the tests pass... appears to hang, probably when calling the cleanup method.

Reply:
Looking again at the current state of this file, the way is now is right. Maybe on Friday I was looking at your original code (not with my suggested change from slack).

Anyway, the returned `func()` for closing seems to be correct as is.